File: jira_db_sync/jira_sync/fields.py
"""フィールド展開 (issues_expanded)

責務:
- jira_fields メタデータから動的にカラム定義と SQL 式を生成
- issues.raw_data から全フィールドを展開して issues_expanded テーブルを作成
- issues_readable ビューの作成
"""

import logging

import pandas as pd

logger = logging.getLogger(__name__)


def _safe_str(val) -> str:
    """NaN/None を空文字に変換"""
    if val is None or (isinstance(val, float) and val != val):
        return ""
    return str(val).lower()


def _to_col_type(schema_type, schema_items) -> str:
    """スキーマタイプから DuckDB カラム型を決定"""
    st = _safe_str(schema_type)
    if st in ("number", "numeric"):
        return "DOUBLE"
    elif st == "datetime":
        return "TIMESTAMPTZ"
    elif st == "date":
        return "DATE"
    elif st in ("array", "any"):
        return "JSON"
    return "VARCHAR"


def _to_select_expr(field_id: str, schema_type, schema_items, schema_system) -> str:
    """フィールドの JSON 展開 SQL 式を生成"""
    st = _safe_str(schema_type)
    fid = field_id

    if st == "user":
        return f"i.raw_data->'fields'->'{fid}'->>'displayName'"
    elif st in ("status", "priority", "resolution", "issuetype", "issuelink",
                "securitylevel", "component", "version"):
        return f"i.raw_data->'fields'->'{fid}'->>'name'"
    elif st in ("option", "option-with-child"):
        return (f"COALESCE(i.raw_data->'fields'->'{fid}'->>'value',"
                f"i.raw_data->'fields'->'{fid}'->>'name')")
    elif st == "array":
        return f"TRY_CAST(i.raw_data->'fields'->'{fid}' AS JSON)"
    elif st == "number":
        return f"TRY_CAST(i.raw_data->'fields'->>'{fid}' AS DOUBLE)"
    elif st in ("datetime", "date"):
        return f"TRY_CAST(i.raw_data->'fields'->>'{fid}' AS TIMESTAMP)"
    elif st in ("progress", "any"):
        return f"TRY_CAST(i.raw_data->'fields'->'{fid}' AS JSON)"
    elif st == "string":
        return f"i.raw_data->'fields'->>'{fid}'"
    else:
        return (f"COALESCE(i.raw_data->'fields'->'{fid}'->>'name',"
                f"i.raw_data->'fields'->'{fid}'->>'value',"
                f"i.raw_data->'fields'->'{fid}'->>'displayName',"
                f"i.raw_data->'fields'->>'{fid}')")


class FieldExpander:
    """jira_fields メタデータに基づいて issues_expanded テーブルを構築"""

    def __init__(self, conn, fields_df: pd.DataFrame):
        self.conn = conn
        self.fields_df = fields_df

    def expand(self, project_key: str) -> dict:
        """issues.raw_data からフィールドを展開して issues_expanded に投入

        Returns:
            {"expanded": 件数, "columns": カラム数, "custom_fields": カスタムフィールド数}
        """
        target_fields = self.fields_df[
            self.fields_df["navigable"] == True
        ].to_dict("records")

        # 固定カラム
        select_parts = [
            "i.id",
            "i.project_id",
            'COALESCE(i.raw_data->>\'key\', i."key") AS issue_key',
        ]
        col_defs = [
            ("id", "VARCHAR PRIMARY KEY"),
            ("project_id", "VARCHAR NOT NULL"),
            ("issue_key", "VARCHAR NOT NULL"),
        ]
        processed = {"id", "project_id", "issue_key"}

        # メタデータから動的にカラムを生成
        custom_count = 0
        for f in target_fields:
            fid = f["id"]
            col = fid.lower().replace("-", "_").replace(".", "_")
            if col in processed:
                continue
            processed.add(col)

            col_type = _to_col_type(f.get("schema_type"), f.get("schema_items"))
            expr = _to_select_expr(
                fid, f.get("schema_type"), f.get("schema_items"), f.get("schema_system")
            )
            select_parts.append(f'{expr} AS "{col}"')
            col_defs.append((col, col_type))
            if f.get("custom"):
                custom_count += 1

        select_parts.append("CURRENT_TIMESTAMP AS synced_at")
        col_defs.append(("synced_at", "TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP"))

        # テーブル再作成
        self.conn.execute("DROP TABLE IF EXISTS issues_expanded")
        create_cols = ", ".join(f'"{c}" {t}' for c, t in col_defs)
        self.conn.execute(f"CREATE TABLE issues_expanded ({create_cols})")
        self.conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_expanded_project ON issues_expanded(project_id)"
        )
        self.conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_expanded_key ON issues_expanded(issue_key)"
        )

        # データ投入
        select_sql = ",\n        ".join(select_parts)
        self.conn.execute(f"""
            INSERT INTO issues_expanded
            SELECT {select_sql}
            FROM issues i
            WHERE i."key" LIKE '{project_key}-%'
        """)

        # issues_readable ビュー
        self._create_readable_view(col_defs)

        expanded = self.conn.execute("SELECT COUNT(*) FROM issues_expanded").fetchone()[0]
        logger.info(
            "Expanded %d issues into %d columns (%d custom)",
            expanded, len(col_defs), custom_count,
        )
        return {
            "expanded": expanded,
            "columns": len(col_defs),
            "custom_fields": custom_count,
        }

    def _create_readable_view(self, col_defs: list[tuple[str, str]]):
        """issues_readable ビュー（カラム名 → フィールド名）を作成"""
        field_name_map = dict(
            zip(self.fields_df["id"].str.lower(), self.fields_df["name"])
        )
        field_name_map.update({
            "id": "ID", "project_id": "Project ID",
            "issue_key": "Key", "synced_at": "同期日時",
        })
        view_cols = []
        for col, _ in col_defs:
            display = field_name_map.get(col, col).replace('"', '""')
            view_cols.append(f'"{col}" AS "{display}"')
        self.conn.execute(
            f"CREATE OR REPLACE VIEW issues_readable AS SELECT {', '.join(view_cols)} FROM issues_expanded"
        )

File: jd-go/core/fields.go
package core

import (
	"database/sql"
	"fmt"
	"log/slog"
	"strings"
)

func safeStr(s *string) string {
	if s == nil {
		return ""
	}
	return strings.ToLower(*s)
}

func toColType(schemaType, schemaItems *string) string {
	st := safeStr(schemaType)
	switch st {
	case "number", "numeric":
		return "DOUBLE"
	case "datetime":
		return "TIMESTAMPTZ"
	case "date":
		return "DATE"
	case "array", "any":
		return "JSON"
	default:
		return "VARCHAR"
	}
}

func toSelectExpr(fieldID string, schemaType, schemaItems, schemaSystem *string) string {
	st := safeStr(schemaType)
	fid := fieldID

	switch st {
	case "user":
		return fmt.Sprintf("i.raw_data->'fields'->'%s'->>'displayName'", fid)
	case "status", "priority", "resolution", "issuetype", "issuelink",
		"securitylevel", "component", "version":
		return fmt.Sprintf("i.raw_data->'fields'->'%s'->>'name'", fid)
	case "option", "option-with-child":
		return fmt.Sprintf("COALESCE(i.raw_data->'fields'->'%s'->>'value',i.raw_data->'fields'->'%s'->>'name')", fid, fid)
	case "array":
		return fmt.Sprintf("TRY_CAST(i.raw_data->'fields'->'%s' AS JSON)", fid)
	case "number":
		return fmt.Sprintf("TRY_CAST(i.raw_data->'fields'->>'%s' AS DOUBLE)", fid)
	case "datetime", "date":
		return fmt.Sprintf("TRY_CAST(i.raw_data->'fields'->>'%s' AS TIMESTAMP)", fid)
	case "progress", "any":
		return fmt.Sprintf("TRY_CAST(i.raw_data->'fields'->'%s' AS JSON)", fid)
	case "string":
		return fmt.Sprintf("i.raw_data->'fields'->>'%s'", fid)
	default:
		return fmt.Sprintf("COALESCE(i.raw_data->'fields'->'%s'->>'name',i.raw_data->'fields'->'%s'->>'value',i.raw_data->'fields'->'%s'->>'displayName',i.raw_data->'fields'->>'%s')",
			fid, fid, fid, fid)
	}
}

type fieldColDef struct {
	name    string
	colType string
}

// FieldExpander dynamically expands issue fields from raw_data.
type FieldExpander struct {
	db     *sql.DB
	fields []JiraField
}

// NewFieldExpander creates a new FieldExpander.
func NewFieldExpander(db *sql.DB, fields []JiraField) *FieldExpander {
	return &FieldExpander{db: db, fields: fields}
}

// Expand creates the issues_expanded table and issues_readable view.
func (e *FieldExpander) Expand(projectKey string) (ExpandResult, error) {
	// Filter navigable fields
	var targetFields []JiraField
	for _, f := range e.fields {
		if f.Navigable {
			targetFields = append(targetFields, f)
		}
	}

	// Fixed columns
	selectParts := []string{
		"i.id",
		"i.project_id",
		`COALESCE(i.raw_data->>'key', i."key") AS issue_key`,
	}
	colDefs := []fieldColDef{
		{"id", "VARCHAR PRIMARY KEY"},
		{"project_id", "VARCHAR NOT NULL"},
		{"issue_key", "VARCHAR NOT NULL"},
	}
	processed := map[string]bool{"id": true, "project_id": true, "issue_key": true}

	// Dynamic columns from metadata
	customCount := 0
	for _, f := range targetFields {
		col := strings.ToLower(f.ID)
		col = strings.ReplaceAll(col, "-", "_")
		col = strings.ReplaceAll(col, ".", "_")
		if processed[col] {
			continue
		}
		processed[col] = true

		ct := toColType(f.SchemaType, f.SchemaItems)
		expr := toSelectExpr(f.ID, f.SchemaType, f.SchemaItems, f.SchemaSystem)
		selectParts = append(selectParts, fmt.Sprintf(`%s AS "%s"`, expr, col))
		colDefs = append(colDefs, fieldColDef{col, ct})
		if f.Custom {
			customCount++
		}
	}

	selectParts = append(selectParts, "CURRENT_TIMESTAMP AS synced_at")
	colDefs = append(colDefs, fieldColDef{"synced_at", "TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP"})

	// Recreate table
	if _, err := e.db.Exec("DROP TABLE IF EXISTS issues_expanded"); err != nil {
		return ExpandResult{}, fmt.Errorf("drop issues_expanded: %w", err)
	}

	var createCols []string
	for _, cd := range colDefs {
		createCols = append(createCols, fmt.Sprintf(`"%s" %s`, cd.name, cd.colType))
	}
	createSQL := fmt.Sprintf("CREATE TABLE issues_expanded (%s)", strings.Join(createCols, ", "))
	if _, err := e.db.Exec(createSQL); err != nil {
		return ExpandResult{}, fmt.Errorf("create issues_expanded: %w", err)
	}

	e.db.Exec("CREATE INDEX IF NOT EXISTS idx_expanded_project ON issues_expanded(project_id)")
	e.db.Exec("CREATE INDEX IF NOT EXISTS idx_expanded_key ON issues_expanded(issue_key)")

	// Insert data
	selectSQL := strings.Join(selectParts, ",\n        ")
	insertSQL := fmt.Sprintf(`INSERT INTO issues_expanded
		SELECT %s
		FROM issues i
		WHERE i."key" LIKE '%s-%%'`, selectSQL, projectKey)
	if _, err := e.db.Exec(insertSQL); err != nil {
		return ExpandResult{}, fmt.Errorf("insert issues_expanded: %w", err)
	}

	// Create readable view
	e.createReadableView(colDefs)

	var expanded int
	e.db.QueryRow("SELECT COUNT(*) FROM issues_expanded").Scan(&expanded)

	slog.Info("Expanded issues", "count", expanded, "columns", len(colDefs), "custom", customCount)
	return ExpandResult{
		Expanded:     expanded,
		Columns:      len(colDefs),
		CustomFields: customCount,
	}, nil
}

func (e *FieldExpander) createReadableView(colDefs []fieldColDef) {
	fieldNameMap := make(map[string]string)
	for _, f := range e.fields {
		fieldNameMap[strings.ToLower(f.ID)] = f.Name
	}
	fieldNameMap["id"] = "ID"
	fieldNameMap["project_id"] = "Project ID"
	fieldNameMap["issue_key"] = "Key"
	fieldNameMap["synced_at"] = "同期日時"

	var viewCols []string
	for _, cd := range colDefs {
		display := cd.name
		if name, ok := fieldNameMap[cd.name]; ok {
			display = name
		}
		display = strings.ReplaceAll(display, `"`, `""`)
		viewCols = append(viewCols, fmt.Sprintf(`"%s" AS "%s"`, cd.name, display))
	}

	sql := fmt.Sprintf("CREATE OR REPLACE VIEW issues_readable AS SELECT %s FROM issues_expanded",
		strings.Join(viewCols, ", "))
	e.db.Exec(sql)
}
